### DISCUSSION TITLE
ClickHouse driver does not work as expected

### DISCUSSION THREAD

**Discussion by nuclon** [burningalchemist/sql_exporter#629] [Q&A] [answered]
With version 0.16.0, when the connection to clickhouse is used, all queries fail with following error:

`nc=Errorf msg="Error gathering metrics: [from Gatherer #1] [collector=data_check,query=test] invalid INSERT query: select 1 AS test"`

Not sure, why it says "invalid INSERT query", but it seems the problem is somewhere there :)

**Comment by burningalchemist** [comment, marked as answer]
Hi @nuclon, could you please share your configuration (redacted)?

UPD: sql_exporter doesn't contain any db-specific code, so `invalid INSERT query` comes from the Clickhouse driver itself, the only occurrence found is [here](https://github.com/ClickHouse/clickhouse-go/blob/fcf5831835335a86078abfb224fdcc33ab4bddc0/batch.go#L39), which is related to the prepared statements ([PrepareContext](https://github.com/ClickHouse/clickhouse-go/blob/fcf5831835335a86078abfb224fdcc33ab4bddc0/clickhouse_std.go#L351) function). This won't work in the case of sql_exporter. So please make sure you use unprepared statements as stated in the docs:

<img width="886" alt="image" src="https://github.com/user-attachments/assets/a3fdb156-9bbe-42ba-86b5-d0e64151b80d">

**Comment by burningalchemist** [comment]
I'll convert it to discussion at the moment, you may also find related Clickhouse discussions [here](https://github.com/burningalchemist/sql_exporter/discussions?discussions_q=clickhouse+is%3Aclosed).

**Comment by nuclon** [comment]
Thanks for pointing to the documentation.
Adding "no_prepared_statement: true" to the metric config helped.
